File: 0130_Surrounded Regions/SurroundedRegions.cs
public class Solution {
    public void Solve(char[][] board) {
        if(board == null || board.Length == 0) return ;
        var m = board.Length;
        var n = board[0].Length;

        for(int i=0;i<m;i++){
            DFS(board, m, n, i, 0); 
            DFS(board, m, n, i, n-1);
        }

        for(int i=0;i<n;i++){
            DFS(board, m, n, 0, i);
            DFS(board, m, n, m-1, i);
        }

        for(int i=0;i<m;i++){
            for(int j=0;j<n;j++){
                if(board[i][j] == 'O') board[i][j] = 'X';
                if(board[i][j] == 'K') board[i][j] = 'O';
            }
        }
    }

    private void DFS(char[][] board,int m, int n, int x, int y){
        if(x < 0 || y < 0 || x >= m || y >= n || board[x][y] != 'O') return;
        board[x][y] = 'K';
        DFS(board, m, n, x-1, y);
        DFS(board, m, n, x, y-1);
        DFS(board, m, n, x+1, y);
        DFS(board, m, n, x, y+1);
    }
}

File: 0130_Surrounded Regions/SurroundedRegions.java
class Solution {
    public void solve(char[][] board) {
        int m = board.length, n = board[0].length;
        int[] dirs = new int[]{0,1,0,-1,0};
        for(int i=0;i<n;i++) {
            if(board[0][i] == 'O') dfs(board, 0, i, dirs);
            if(board[m-1][i] == 'O') dfs(board, m-1, i, dirs);
        }
        
        for(int i=0;i<m;i++) {
            if(board[i][0] == 'O') dfs(board, i, 0, dirs);
            if(board[i][n-1] == 'O') dfs(board, i, n-1, dirs);
        }
        
        for(int i=0;i<m;i++) {
            for(int j=0;j<n;j++) {
                if(board[i][j] == 'O') board[i][j] = 'X';
                if(board[i][j] == 'M') board[i][j] = 'O';
            }
        }
    }
    
    private void dfs(char[][] board, int x , int y, int[] dirs) {
        if(x < 0 || y < 0 || x == board.length || y == board[x].length || board[x][y] != 'O') return;
        board[x][y] = 'M';
        
        for(int i=0;i<4;i++) {
            int dx = x + dirs[i];
            int dy = y + dirs[i+1];
            dfs(board, dx, dy, dirs);
        }
    }
}